Refactor workflow to extract processing with submission

bdc34 opened the pull request:
This is the direction I was going with the change to workflow. 

It is mainly moving the workfow into a WorkfowDefinition and having a separate WorkflowProcessor that handles where a submission is in a workflow.

This is a early preview PR. After this I'd test the workflows and integrate with routes and flow_control.py.

ARXIVNG-2886

bdc34 committed:
Refactor workflow to extract processing with submission

ARXIVNG-2886

bdc34 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+from arxiv.submission.domain import Submission
```
I split the conditions on the Workflow stages out for reuse and testing.

DavidLFielding replied:
Seems reasonable.

bdc34 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+        submission.source_content.source_format != SubmissionContent.Format.TEX
+        and
+        submission.source_content.source_format != SubmissionContent.Format.POSTSCRIPT)
+    return has_valid_content(submission) and \
```
Here is the reuse. Not as much as I was hoping.

bdc34 commented on submit/workflow/processor.py:
```diff
@@ -0,0 +1,82 @@
+"""Defines submission stages and workflows supported by this UI."""
```
This is a new Class that handles where a submission is within a WorkflowDefinition.

bdc34 commented on submit/workflow/test_new_submission.py:
```diff
@@ -0,0 +1,26 @@
+        submission = mock.MagicMock(submission_id=1234,
+                                    is_finalized=False,
+                                    submitter_contact_verified=False)        
+        nswfps = processor.WorkflowProcessor(
```
This shows how the workflow is used.

DavidLFielding replied:
Test is helpful. Hopefully we'll have more complicated tests in future though this depends on what gets decided in terms of locking down workflow. I'm more in favor of allowing user to select order of steps they complete while disabling step transitions that are invalid (ie., process without uploading files)

DavidLFielding reviewed (approved):
Looks fine to me. Mostly reorganization/refactoring at this point. I'm not clear on expected reuse of conditionals. I can't wait until this is wired up so we can move forward tightening up workflow and eliminating invalid transitions.

mhl10 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+from arxiv.submission.domain import Submission
+
+def contact_verified(submission: Submission) -> bool:
```
```suggestion
def is_contact_verified(submission: Submission) -> bool:
```

mhl10 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+    return submission.submitter_contact_verified is True
+
+
+def authorship_indicated(submission: Submission) -> bool:
```
```suggestion
def is_authorship_indicated(submission: Submission) -> bool:
```

mhl10 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+
+
+
+def policy_accepted(submission: Submission) -> bool:
```
```suggestion
def is_policy_accepted(submission: Submission) -> bool:
```

mhl10 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+
+    
+
+def source_processed(submission: Submission) -> bool:
```
```suggestion
def is_source_processed(submission: Submission) -> bool:
```

bdc34 committed:
Update submit/workflow/conditions.py

Update name of condition functions.

Co-Authored-By: Martin Lessmeister <mhl10@cornell.edu>

mhl10 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+            (submission.is_source_processed or non_processing_content)
+
+
+def metadata_complete(submission: Submission) -> bool:
```
```suggestion
def is_metadata_complete(submission: Submission) -> bool:
```

mhl10 commented on submit/workflow/conditions.py:
```diff
@@ -0,0 +1,68 @@
+            and submission.metadata.abstract is not None
+            and submission.metadata.authors_display is not None)
+
+def opt_metadata_complete(submission: Submission) -> bool:
```
```suggestion
def is_opt_metadata_complete(submission: Submission) -> bool:
```

bdc34 committed:
Changes Workflows from Classes to instances of WorkflowDefinitions

Also I've added a __getitem__ to WorkflowDefinition that handles
Stage classes, stage class names, instances of Stages, stage labels,
and stage endpoints.

This is to make it easy to find stages in workflows.

ARXIVNG-2886

mhl10 commented on submit/workflow/__init__.py:
```diff
@@ -0,0 +1,97 @@
+from arxiv.submission.domain import Submission
+from dataclasses import dataclass
+
+from . import stages
+from .stages import Stage
```
absolute imports may be preferable for clarity

bdc34 replied:
I was thinking that this was a place  where the stages are closely related to the workflows.

bdc34 committed:
Merge branch 'ARXIVNG-2886' of github.com:arXiv/arxiv-submission-ui into ARXIVNG-2886

mhl10 commented on submit/workflow/processor.py:
```diff
@@ -0,0 +1,82 @@
+    submission: Submission
+    seen: Dict[str, bool] = field(default_factory=dict)
+
+    def complete(self) -> bool:
```
```suggestion
    def is_complete(self) -> bool:
```

mhl10 commented on submit/workflow/processor.py:
```diff
@@ -0,0 +1,82 @@
+                return stage
+        return None
+
+    def is_complete(self, stage: Optional[Stage]) -> bool:
```
```suggestion
    def is_stage_complete(self, stage: Optional[Stage]) -> bool:
```

bdc34 committed:
Create a new submission with the event system in workflow tests

This commit includes an example of how to create a submission from
the event modules.

ARXIVNG-2886

bdc34 committed:
Bug fixes to workflow classes.

Changes some workflow.conditions function names to predicate style
names.

Fixes bug with _seen_key() of WorkflowProcessor.

Fixes bug due to class to instance changes of workflows.

bdc34 committed:
Changing is_complete on WorkflowProcessor

Renaming WorkflowProcessor.complete() to is_complete()
Removing WorkflowProcessor.is_complete(stage) in favor of just
using stage.is_complete().

bdc34 committed:
Adding workflow testing for all steps of a new submission

Adds testing of the workflow of a basic new submission.

ARIXVNG-2886

bdc34 commented:
Thanks for you feedback. I've incorporated some of your ideas. 

I'm going to close this PR and open a new one once I have it wired into the flask app.
